io/teca_cf_writer.cxx

```diff
@@ -0,0 +1,649 @@
+        const std::string &t_variable, const teca_metadata &array_attributes,
+        const std::vector<const_p_teca_array_collection> &arrays);
+};
+
+// --------------------------------------------------------------------------
+int teca_cf_writer_internals::write(const std::string &file_name, int mode,
+    const std::vector<long> &request_ids, const const_p_teca_variant_array &x,
+    const const_p_teca_variant_array &y, const const_p_teca_variant_array &z,
+    const const_p_teca_variant_array &t, const std::string &x_variable,
+    const std::string &y_variable, const std::string &z_variable,
+    const std::string &t_variable, const teca_metadata &array_attributes,
+    const std::vector<const_p_teca_array_collection> &arrays)
+{
+    (void) request_ids;
+
+    int ierr = NC_NOERR;
+
+    // create the output file
+    teca_netcdf_util::netcdf_handle fh;
+    if (fh.create(file_name.c_str(), mode))
+    {
+        TECA_ERROR("failed to create file \"" << file_name << "\"")
+        return -1;
+    }
+
+    // files are always written in 4D. at least one of the coordinates must
+    // have data for the others construct a length one array containing zeros
+    const_p_teca_variant_array coord_array
+        = x ? x : y ? y : z ? z : t ? t : nullptr;
+
+    if (!coord_array)
+    {
+        TECA_ERROR("invalid coordinates")
+        return -1;
+    }
+
+    const_p_teca_variant_array coord_arrays[4];
+    std::string coord_array_names[4];
+
+    // the cf reader always creates 4D data, but some other tools choke
+    // on it, notably ParView. All dimensions of 1 are safe to skip, unless
+    // we are writing a variable with 1 value.
+    unsigned long skip_dim_of_1 = (x && x->size() > 1 ? 1 : 0) +
+        (y && y->size() > 1 ? 1 : 0) + (z  && z->size() > 1 ? 1 : 0);
+
+    int n_dims = 0;
+    if (t)
+    {
+        coord_arrays[n_dims] = t;
+        coord_array_names[n_dims] = t_variable.empty() ? "time" : t_variable;
+        ++n_dims;
+    }
+    if (z)
+    {
+        if (!skip_dim_of_1 || z->size() > 1)
+        {
+            coord_arrays[n_dims] = z;
+            coord_array_names[n_dims] = z_variable.empty() ? "z" : z_variable;
+            ++n_dims;
+        }
+    }
+    if (y)
+    {
+        if (!skip_dim_of_1 || y->size() > 1)
+        {
+            coord_arrays[n_dims] = y;
+            coord_array_names[n_dims] = y_variable.empty() ? "y" : y_variable;
+            ++n_dims;
+        }
+    }
+    if (x)
+    {
+        if (!skip_dim_of_1 || x->size() > 1)
+        {
+            coord_arrays[n_dims] = x;
+            coord_array_names[n_dims] = x_variable.empty() ? "x" : x_variable;
+            ++n_dims;
+        }
+    }
+
+    // dictionary of names to ncids
+    int dim_ids[4] = {-1};
+    size_t dims[4] = {0, 0, 0, 0};
+    std::map<std::string, int> var_ids;
+
+    for (int i = 0; i < n_dims; ++i)
+    {
+        // save the size of the dimension
+        dims[i] = coord_arrays[i]->size();
```

Comment:
@burlen - based on what I'm seeing here, `coord_arrays[i]->size()` will always return an actually array size, right?  In order to set dimension `i` as an unlimited dimension (which is almost always going to be 'time' in our case), we would need `dims[i] = NC_UNLIMITED`, which is a special value passed for the `len` argument when calling `nc_def_dim()`.  There are some other considerations too, like the order in which an unlimited dimension is defined relative to others, depending on whether the outputted file is a classic NetCDF file or an HDF-5 one: see https://www.unidata.ucar.edu/software/netcdf/docs/unlimited_dims.html and https://www.unidata.ucar.edu/software/netcdf/netcdf-4/newdocs/netcdf-c/nc_005fdef_005fdim.html

We need some sort of mechanism to indicate that a dimension is an 'unlimited' dimension -- it will almost always be the time dimension in our case.  

Two reasonable options for handling this come to mind:
 1. via metadata, e.g. a special attribute like `array_atts["is_unlimited"] = true`.
 2. via an extra argument to `teca_cf_writer_internals::write()`: e.g., `bool t_is_unlimited`

I think that (2) is the simplest (would involve far less plumbing).  (1) would allow generalizability to have multiple unlimited dims, but that would require much more code to implement.  I don't actually think (1) is actually worth the effort, since we don't have an immediate use case for that and since I've never actually seen netCDF files with more than one unlimited dim (or one where a dimension other than time is the unlimited dimension).  But I figured I'd throw it out as an option that might be worth discussing.

Either way, I think a good default would be to assume that any dimension named 'time' should be an unlimited dimension, since this is the prevailing practice in netCDF files used in climate/atmospheric science.

Reply:
Sorry that I did not see this when you wrote it!

I agree and was half expecting to need to add this feature. I wold have done it in the first place I had had more experience writing netcdf files. It turns out it is an easy change.

I took the approach you described in 2, with the slight modification that if the flag is set then the slowest varying dimension is made unlimited.  unlimited dims are on by default.

I added some notes about performance implications, and also some documentation about how to get optimal through put.

Reply:
Looks perfect, and nice job on this!